src/paths/steiner.cpp

```diff
@@ -0,0 +1,650 @@
+	}
+	return answer;
+}
+
+/*
+ * Finding number of combinations nCr
+ * Used to determine number of elements to be scanned in DP table
+ * for a particular subset during generation of Steiner Tree.
+ *
+ */
+
+igraph_integer_t Combination(igraph_integer_t n, igraph_integer_t r)
+{
+	return factorial(n) / (factorial(n - r) * factorial(r));
+}
+
+/* \function  findMinimumK
+ *
+ *
+ *  Iterate through a particular row of the DP table
+ *  find the value of column where distance (q,k) + distance(Sk(D))
+ *  where D is a subset. This signifies a vertex that is part
+ *  of minimum path from vertex q to D
+ *  \param dp_cache The DP table.
+ *
+ *  \param indexD Index of the subset D.
+ *
+ *  \param q vertex from who minimum path needs tp be calculated
+ */
+
+igraph_integer_t findMinimumK(igraph_matrix_t *dp_cache, igraph_integer_t indexD, igraph_integer_t q)
+{
+
+	igraph_integer_t min_col_num = -1;
+	igraph_integer_t min_sum_for_col;
+
+	for (igraph_integer_t i = 0; i < dp_cache->ncol; i++)
+	{
+		if (q != i)
+		{
+
+			if (min_col_num == -1)
+			{
+				min_col_num = i;
+				min_sum_for_col = MATRIX(*dp_cache, q, i) + MATRIX(*dp_cache, indexD, i);
+			}
+			else if (MATRIX(*dp_cache, q, i) + MATRIX(*dp_cache, indexD, i) < min_sum_for_col)
+			{
+				min_col_num = i;
+				min_sum_for_col = MATRIX(*dp_cache, q, i) + MATRIX(*dp_cache, indexD, i);
+			}
+		}
+	}
+
+	return min_col_num;
+}
+
+/* \function generate_steiner_tree_appx()
+ *
+ *   Generation of the Steiner Tree based on calculation of minimum distances in DP table.
+ *
+ *  \param graph The graph object.
+ *  \param weights The edge weights. All edge weights must be
+ *                 non-negative. Additionally, no edge weight may be NaN.
+ *
+ *  \param dp_cache The DP table.
+ *
+ *  \param indexD The index of subset D in DP table.
+ *
+ *  \param q The vertex that was remmoved from steiner terminals.
+ *
+ * \param mode How to determine the local neighborhood of each vertex
+ *  in directed graphs. Ignored in undirected graphs.
+ * \clist
+ *         \cli IGRAPH_ALL
+ *              take both in- and out-neighbours;
+ *              this is a reasonable default for high-level interfaces.
+ *         \cli IGRAPH_OUT
+ *              take only out-neighbours
+ *         \cli IGRAPH_IN
+ *              take only in-neighbours
+ *         \endclist
+ *
+ *  \param vectorlist_all The vector to capture vertices in resultant Steiner Tree.
+ *
+ *  \param edgelist_all The vector to capture edges in resultant Steiner Tree.
+ *
+ */
+
+igraph_error_t generate_steiner_tree_appx(const igraph_t *graph, const igraph_vector_t *weights,
+										  igraph_matrix_t *dp_cache, igraph_integer_t indexD, igraph_integer_t q, igraph_vector_int_t *vectorlist_all, igraph_vector_int_t *edgelist_all)
+{
+	
+	std::set<igraph_integer_t> C = fetchSetsBasedonIndex(indexD);
+
+	// Initially the value of m is the vertex that was removed from Steiner Terminals
+	igraph_integer_t m = q;
+
+	int len = C.size();
```

Comment:
Let's stick to `igraph_integer_t` unless there is a good reason to use `int`. Keep in mind that `int` will be 32-bit on basically all supported platforms while `igraph_integer_t` will be 64-bit on 64-bit systems.